### DISCUSSION TITLE
Remove residual rpm packages after deleting the Requires dependencies

### DISCUSSION THREAD

**Discussion by xiacunshun** [rpm-software-management/rpm#2980] [Q&A]
Generally requests for new features should be opened as a [Discussion](https://github.com/rpm-software-management/rpm/discussions) first. There feedback from the community can be collected and possible solutions can be discussed. Consider maintaining a design outline in the top post and update it as the discussion progresses. After a plan of action has solidified an issue for actually implementing it can be created.

**Is your feature request related to a problem? Please describe.**

When I remove the unnecessary 'Requires:' of one package,  I found that the residual package which is no more required can not br removed when upgrade. I tried 'Obsoletes', but I found that it affects other packages that requires it. For example, in package A, if 'Requires B' is removed and 'Obsoletes B' is added, then another package C that 'Requires B' cannot be installed.

**Describe the solution you'd like**

Something like 'Obsoletes' that can only remove unneccesary requires of current package.

**Describe alternatives you've considered**

**Additional context**

**Comment by xiacunshun** [comment]
[root@linux ~]# dnf install python3-cheroot-10.0.0-2.ocs23.noarch.rpm
Last metadata expiration check: 3:22:05 ago on Tue Mar 12 16:17:04 2024.
python3-cheroot-10.0.0-2.ocs23.noarch.rpm          3.8 MB/s | 217 kB     00:00    
Error: 
 Problem: problem with installed package python3-conda-23.9.0-1.ocs23.noarch
  - package python3-conda-23.9.0-1.ocs23.noarch from @System requires python3.11dist(pyopenssl) >= 16.2, but none of the providers can be installed
  - package python3-conda-23.9.0-1.ocs23.noarch from EPOL requires python3.11dist(pyopenssl) >= 16.2, but none of the providers can be installed
  - package python3-cheroot-10.0.0-2.ocs23.noarch from @commandline obsoletes python3-pyOpenSSL <= 23.2.0 provided by python3-pyOpenSSL-23.2.0-2.ocs23.noarch from @System
  - package python3-cheroot-10.0.0-2.ocs23.noarch from @commandline obsoletes python3-pyOpenSSL <= 23.2.0 provided by python3-pyOpenSSL-23.2.0-2.ocs23.noarch from EPOL
  - conflicting requests
(try to add '--skip-broken' to skip uninstallable packages or '--nobest' to use not only best candidate packages)

**Comment by pmatilai** [comment]
> When I remove the unnecessary 'Requires:' of one package, I found that the residual package which is no more required can not br removed when upgrade. I tried 'Obsoletes', but I found that it affects other packages that requires it.

Impossible to give any detailed response because there's almost no info provided on what was done and so on, but: if a package cannot be removed after removing another one, then it sounds to me like that package is missing a dependency on the removed one. Also, if you used 'rpm --nodeps' (the most common way of ending up with broken dependencies) then you get to keep all the pieces.

It's also not a feature request of any kind. Questions like this is what the Q&A department in the discussions is for.

**Reply by xiacunshun** [reply]
I apologize if my previous description was not very precise. What I meant to say was that if A originally depends on B, and we remove the "Requires B" from the spec, the next time we upgrade A using "dnf upgrade" or "rpm -U", it will not simultaneously remove B. This is different from the effect of directly using "dnf remove A", because the latter will remove unused dependencies B.
